azbus/disposition.go

```diff
@@ -0,0 +1,147 @@
+}
+
+// NB: ALL disposition methods return nil so they can be used in return statements
+
+// Abandon abandons message. This function is not used but is present for consistency.
+func (r *Receiver) Abandon(ctx context.Context, err error, msg *ReceivedMessage) error {
+	ctx = context.WithoutCancel(ctx)
+	log := r.log.FromContext(ctx)
+	defer log.Close()
+
+	span, ctx := tracing.StartSpanFromContext(ctx, "Message.Abandon")
+	defer span.Finish()
+	log.Infof("Abandon Message on DeliveryCount %d: %v", msg.DeliveryCount, err)
+	err1 := r.receiver.AbandonMessage(ctx, msg, nil)
+	if err1 != nil {
+		azerr := fmt.Errorf("Abandon Message failure: %w", NewAzbusError(err1))
+		log.Infof("%s", azerr)
+	}
+	return nil
+}
+
+// Reschedule handles when a message should be deferred at a later time. There are a
+// number of ways of doing this but it turns out that simply not doing anything causes
+// azservicebus to resubmit the message 1 minute later. We keep the function signature with
+// unused arguments for consistency and in case we need to implement more sophisticated
+// algorithms in future.
+func (r *Receiver) Reschedule(ctx context.Context, err error, msg *ReceivedMessage) error {
+	ctx = context.WithoutCancel(ctx)
+	log := r.log.FromContext(ctx)
+	defer log.Close()
+
+	span, _ := tracing.StartSpanFromContext(ctx, "Message.Reschedule")
+	defer span.Finish()
+	log.Infof("Reschedule Message on DeliveryCount %d: %v", msg.DeliveryCount, err)
+	return nil
+}
+
+// DeadLetter explicitly deadletters a message.
+func (r *Receiver) DeadLetter(ctx context.Context, err error, msg *ReceivedMessage) error {
+	ctx = context.WithoutCancel(ctx)
+	log := r.log.FromContext(ctx)
+	defer log.Close()
+
+	span, ctx := tracing.StartSpanFromContext(ctx, "Message.DeadLetter")
+	defer span.Finish()
+	log.Infof("DeadLetter Message: %v", err)
+	options := azservicebus.DeadLetterOptions{
+		Reason: to.Ptr(err.Error()),
+	}
+	err1 := r.receiver.DeadLetterMessage(ctx, msg, &options)
+	if err1 != nil {
+		azerr := fmt.Errorf("DeadLetter Message failure: %w", NewAzbusError(err1))
+		log.Infof("%s", azerr)
+	}
+	return nil
+}
+
+func (r *Receiver) Complete(ctx context.Context, msg *ReceivedMessage) error {
+	ctx = context.WithoutCancel(ctx)
+	log := r.log.FromContext(ctx)
+	defer log.Close()
+
+	span, _ := tracing.StartSpanFromContext(ctx, "Message.Complete")
+	defer span.Finish()
+
+	log.Infof("Complete Message")
+
+	err := r.receiver.CompleteMessage(ctx, msg, nil)
+	if err != nil {
+		// If the completion fails then the message will get rescheduled, but it's effect will
+		// have been made, so we could get duplication issues.
+		azerr := fmt.Errorf("Complete: failed to settle message: %w", NewAzbusError(err))
+		log.Infof("%s", azerr)
+	}
+	return nil
+}
+
+// like Complete but an error may indicate that this will be tried again. Useful for polling type operations
+func (r *Receiver) parallelComplete(ctx context.Context, err error, msg *ReceivedMessage) error {
```

Comment:
this is exactly the same as complete, except we pass an error in that we log if its not nil?

Can we not just log any caller error in the caller? that is our general pattern:

```
if err != nil {
  log.Infof("oops an error occured: %v", err)
  receiver.Complete(ctx, msg)
}
```

Reply:
This was done to accommodate the simplehashAnchor failure - 
we have to keep the old Complete() method because other services still use it. 

This is following the same pattern as Reschedule(),Abandon() and Deadletter() which log the error reported.

Reply:
ehhh bit of a chicken and egg then?

Can we at least call into `complete` then instead?

```
func parallelComplete(ctx, err, msg) {
  if err != nil {
     log.Infof("message: %v, err")
  }

  return Complete(ctx, msg)
}
```

that way we are not duplicating code? It also confused me that this is meant for parallel consumption, but the only difference is a log message?